Add additional hints and clarifications to PSE5 Network Delay Time (#11)
Ada-Developers-Academy/cs-fun-interview-practice: 05/implement-network-delay.instructor.md

2. Do the nodes start at 0 or 1?

   The problem defines the nodes as being labeled from 1 to n, and the examples support this, showing nodes starting at 1. So we will work with this as a requirement.

   This is useful to consider as it impacts how we relate our node data (1-based) to supplemental data (such as the total cost list, or potentially a previous list if we were asked to return path data). If we use a list to track the total cost to reach each node, we'll need to be careful about how we index into the list, either adjusting the node value to be 0-based or inserting a dummy value to effectively make the list 1-based. In the latter case, we would also need to ensure that the dummy value didn't interfere with our logic for determining the maximum cost.

3. Will the graph ever be empty? What should the result be?

>    Since the data is being presented as a list of edges, there is a dta configuration that might feel like a natural representation of an empty graph (a graph having neither nodes nor edges): an empty list. However, technically, we can really only tell that there are no edges in such a graph. There could be any number of nodes, simple that share no edges. So we can't really tell whether the graph is empty or not. That is, we cannot tell the difference between a graph with no nodes and a graph with potentially many unconnected nodes.

Reviewer: - typo: "dta" should be "data"
- should "There could be any number of nodes, simple that share no edges" be "simply"?

Author: ```suggestion
   Since the data is being presented as a list of edges, there is a data configuration that might feel like a natural representation of an empty graph (a graph having neither nodes nor edges): an empty list. However, technically, we can really only tell that there are no edges in such a graph. There could be any number of nodes, simply that share no edges. So we can't really tell whether the graph is empty or not. That is, we cannot tell the difference between a graph with no nodes and a graph with potentially many unconnected nodes.
```

>    However, we are also given the count of nodes int he graph as a separate input, which does allow us to differentiate. A truly empty graph will have an empty list of edges and a count of zero nodes. A graph with no edges will have an empty list of edges and a count of some non-zero number of nodes.

Reviewer: ```suggestion
   However, we are also given the count of nodes in the graph as a separate input, which does allow us to differentiate. A truly empty graph will have an empty list of edges and a count of zero nodes. A graph with no edges will have an empty list of edges and a count of some non-zero number of nodes.
```

Author: What can I say? My fingers are trained to declare integer variables! I make this typo ALL THE TIME. 😆

            # if the neighbor has already been visited, continue
            if neighbor in visited:
                continue
            # calculate the total time to travel to the neighbor
            total_time = time + neighbor_time
            # if the total time is less than the previous time stored to travel to the neighbor
            if total_time < time_needed[neighbor - 1]:
                # store the total time as the time needed to travel to the neighbor
                time_needed[neighbor - 1] = total_time
                # push onto heap to be visited later
                heappush(heap, (total_time, neighbor))

    # return the max time in time_needed list if all of the nodes have been visited, otherwise return -1
    return max(time_needed) if len(visited) == n else -1
```

As written, this approach includes a few checks to help tamp down the growth of the heap. For example, if a node has already been visited, we can skip it. If the total time to reach a node is greater than the time we already have stored, we can skip it. These checks are not strictly necessary, but they do help to reduce the size of the heap and the number of operations we need to perform.

> However, without the `decrease_key` operation, we cannot update the priority of a node that is already in the heap. This means that we will have to add duplicate nodes to the heap, in the worst case approaching O(E). Considering a fully connected graph, the first node would itself add N-1 nodes to the heap, the second node would add N-2 nodes to the heap, and so on. It would be possible to construct edge weights such that all other nodes are processed before the N-2 not processed from the outset would remain on the heap until all other nodes have been processed. This would be possible for each of the other N-1 nodes as well. The total size of the heap, while never fully reaching E, would grow with respect to E, making it O(E). Only `decrease_key` or abandoning the use of a heap (with obvious time complexity repercussions) can guarantee to keep the heap size at O(N).

Reviewer: I don't understand this sentence "It would be possible to construct edge weights such that all other nodes are processed before the N-2 not processed from the outset would remain on the heap until all other nodes have been processed."

Are there extra words in here? Or am I drowsy from a big lunch

Author: Key missing word for the flow: that
```suggestion
However, without the `decrease_key` operation, we cannot update the priority of a node that is already in the heap. This means that we will have to add duplicate nodes to the heap, in the worst case approaching O(E). Considering a fully connected graph, the first node would itself add N-1 nodes to the heap, the second node would add N-2 nodes to the heap, and so on. It would be possible to construct edge weights such that all other nodes that are processed before the N-2 not processed from the outset would remain on the heap until all other nodes have been processed. This would be possible for each of the other N-1 nodes as well. The total size of the heap, while never fully reaching E, would grow with respect to E, making it O(E). Only `decrease_key` or abandoning the use of a heap (with obvious time complexity repercussions) can guarantee to keep the heap size at O(N).
```
But I think a little wordsmithing there is in order. Maybe reorder the clauses a bit... Hmm...

Reviewer: Ah, yes the 'that' makes it flow better. But it is a little wordy. Agree about the wordsmithing

Author: ```suggestion
However, without the `decrease_key` operation, we cannot update the priority of a node that is already in the heap. This means that we will have to add duplicate nodes to the heap, in the worst case approaching O(E). Considering a fully connected graph, the first node would itself add N-1 nodes to the heap, the second node would add N-2 nodes to the heap, and so on. It would be possible to construct edge weights such that as we visit each node, all but one of its edge connections would remain unprocessed (that is, on the heap) until after the entire rest of the graph has been visited. The total size of the heap, while never quite fully reaching E, would grow with respect to E, making it O(E). Only `decrease_key` or abandoning the use of a heap (with obvious time complexity repercussions) can guarantee to keep the heap size at O(N).
```

Reviewer: Nice!
